packages/vite/src/node/idResolver.ts

```diff
@@ -0,0 +1,90 @@
+import type { PartialResolvedId } from 'rollup'
+import aliasPlugin from '@rollup/plugin-alias'
+import type { ResolvedConfig } from './config'
+import type { EnvironmentPluginContainer } from './server/pluginContainer'
+import { createEnvironmentPluginContainer } from './server/pluginContainer'
+import { resolvePlugin } from './plugins/resolve'
+import type { InternalResolveOptions } from './plugins/resolve'
+import { getFsUtils } from './fsUtils'
+import type { Environment } from './environment'
+import type { PartialEnvironment } from './baseEnvironment'
+
+export type ResolveIdFn = (
+  environment: PartialEnvironment,
+  id: string,
+  importer?: string,
+  aliasOnly?: boolean,
+) => Promise<string | undefined>
+
+/**
+ * Create an internal resolver to be used in special scenarios, e.g.
+ * optimizer and handling css @imports
+ */
+export function createIdResolver(
+  config: ResolvedConfig,
+  options: Partial<InternalResolveOptions>,
+): ResolveIdFn {
+  const scan = options?.scan
+
+  const pluginContainerMap = new Map<
+    PartialEnvironment,
+    EnvironmentPluginContainer
+  >()
+  async function resolve(
+    environment: PartialEnvironment,
+    id: string,
+    importer?: string,
+  ): Promise<PartialResolvedId | null> {
+    let pluginContainer = pluginContainerMap.get(environment)
+    if (!pluginContainer) {
+      pluginContainer = await createEnvironmentPluginContainer(
+        environment as Environment,
```

Comment:
Maybe we can change the first parameter to accept `PartialEnvironment` instead of type assertion by making the plugin container `PluginContainer<T extends PartialEnvironment>`?

Reply:
`PluginContainer` generally works fine with `Environment`. The way id resolver is implemented is kind of a hack, so I think we can accept the cast. We are using a plugin container only to make it easier to call `resolve` on the Alias and Resolve plugin. And this two plugins are coded in a special way to avoid depending on a real environment. I think the right thing to do here would be to refactor these two plugins to expose an alias and resolve util free functions that we can use without a plugin container or plugins at all. I tried to do it, but it isn't that simple, so maybe we could let this for a future cleanup.

Reply:
I see. Agreed to leaving it for now 👍